Windows/Impl/PaZword.Core/Data/UpgradeService.cs

```diff
@@ -66,10 +72,45 @@ public async Task<(bool updated, UserDataBundle userDataBundle)> MigrateUserData
         /// </summary>
         private UserDataBundle Load(string encryptedData)
         {
-            string jsonData = _encryptionProvider.DecryptString(encryptedData);
+            int lastColonPosition = encryptedData.LastIndexOf(':');
+            string jsonData = _encryptionProvider.DecryptString(encryptedData.Substring(lastColonPosition + 1));
             return _serializationProvider.DeserializeObject<UserDataBundle>(jsonData);
         }
 
+        /// <summary>
+        /// Migrates from version 1 to 2.
+        /// </summary>
+        private async Task<UserDataBundle> UpgradeToVersion2Async(string encryptedData)
+        {
+            // Version 1 can be loaded as version 2.
+            // The difference between 1 and 2 is a vulnerability in the encryption engine.
+            // To fix it, the data should be decrypted and re-encrypted.
+            _logger.LogEvent(Version1ToVersion2EventName, string.Empty);
+            const int oldVersion = 1;
+            const int newVersion = 2;
+
+            string jsonData = _encryptionProvider.DecryptString(encryptedData);
+            UserDataBundle userDataBundle = _serializationProvider.DeserializeObject<UserDataBundle>(jsonData);
+
+            var tasks = new List<Task>();
+            for (int i = 0; i < userDataBundle.Accounts.Count; i++)
+            {
+                Account account = userDataBundle.Accounts[i];
+                for (int j = 0; j < account.Data.Count; j++)
+                {
+                    AccountData accountData = account.Data[j];
+                    if (accountData is IUpgradableAccountData upgradableAccountData)
+                    {
+                        tasks.Add(upgradableAccountData.UpgradeAsync(oldVersion, newVersion));
```

Comment:
Don't know if it's a concern here, but if you wanted to limit concurrency, TPL dataflow is a very nice but surprisingly little-known library https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/dataflow-task-parallel-library